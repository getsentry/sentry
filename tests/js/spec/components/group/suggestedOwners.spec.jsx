import React from 'react';
import {mount} from 'enzyme';
import SuggestedOwners from 'app/components/group/suggestedOwners';
import MemberListStore from 'app/stores/memberListStore';

describe('SuggestedOwners', function() {
  const event = TestStubs.Event();
  const USER = {
    id: '1',
    name: 'Jane Doe',
    email: 'janedoe@example.com',
  };
  const org = TestStubs.Organization();
  const project = TestStubs.Project();
  const group = TestStubs.Group();
  const endpoint = `/projects/${org.slug}/${project.slug}/events/${event.id}`;

  beforeEach(function() {
    MemberListStore.loadInitialData([USER]);
    MockApiClient.addMockResponse({
      url: `${endpoint}/committers/`,
      body: {committers: []},
    });
    MockApiClient.addMockResponse({
      url: `${endpoint}/owners/`,
      body: {
        owners: [],
        rule: [],
      },
    });
  });

  afterEach(function() {
    MockApiClient.clearMockResponses();
  });

  it('should show owners when enabled', function() {
    MockApiClient.addMockResponse({
      url: `${endpoint}/owners/`,
      body: {
        owners: [
          {
            type: 'user',
            id: '1',
            name: 'Jane Doe',
          },
        ],
        rule: ['path', 'sentry/tagstore/*'],
      },
    });
    let wrapper = mount(
      <SuggestedOwners
        event={event}
        group={group}
        project={project}
        org={TestStubs.Organization({
          features: new Set(['code-owners']),
        })}
      />,
      TestStubs.routerContext([{project, group}])
    );

    expect(wrapper.find('.avatar-grid-item')).toHaveLength(1);

    expect(wrapper.find('.avatar-grid-item')).toMatchSnapshot();
  });

  it('should not show owners committers without featureflag', function() {
    MockApiClient.addMockResponse({
      url: `${endpoint}/owners/`,
      body: {
        owners: [
          {
            type: 'user',
            id: '1',
            name: 'Jane Doe',
          },
        ],
        rule: ['path', 'sentry/tagstore/*'],
      },
    });
    let wrapper = mount(
      <SuggestedOwners event={event} group={group} project={project} org={org} />,
      TestStubs.routerContext([{project, group}])
    );
    expect(wrapper.find('.avatar-grid-item')).toHaveLength(0);
  });

  it('always displays Suggest Owners box', function() {
    let wrapper = mount(
      <SuggestedOwners event={event} group={group} project={project} org={org} />,
      TestStubs.routerContext([{project, group}])
    );
    expect(wrapper.find('h6')).toHaveLength(1);
  });

  it('does not have Add Rule button without feature-flag', function() {
    let wrapper = mount(
      <SuggestedOwners event={event} group={group} project={project} org={org} />,
      TestStubs.routerContext([{project, group}])
    );
    expect(wrapper.find('Tooltip Button')).toHaveLength(0);

    wrapper = mount(
      <SuggestedOwners
        event={event}
        group={group}
        project={project}
        org={TestStubs.Organization({
          features: new Set(['code-owners']),
        })}
      />,
      TestStubs.routerContext([{project, group}])
    );
    expect(wrapper.find('Tooltip Button')).toHaveLength(1);
  });

  it('does allow Add Rule button to be clicked without `project:write`', function() {
    let wrapper = mount(
      <SuggestedOwners
        event={event}
        group={group}
        project={project}
        org={TestStubs.Organization({
          features: new Set(['code-owners']),
        })}
      />,
      TestStubs.routerContext([{project, group}])
    );
    expect(wrapper.find('Tooltip Button')).toHaveLength(1);
    expect(wrapper.find('Tooltip').prop('disabled')).toBe(true);

    wrapper = mount(
      <SuggestedOwners
        event={event}
        group={group}
        project={project}
        org={TestStubs.Organization({
          access: [],
          features: new Set(['code-owners']),
        })}
      />,
      TestStubs.routerContext([{project, group}])
    );
    expect(wrapper.find('Tooltip Button')).toHaveLength(1);
    // Button should be disabled, tooltip should be enabled
    expect(wrapper.find('Tooltip').prop('disabled')).toBe(false);
  });
});
