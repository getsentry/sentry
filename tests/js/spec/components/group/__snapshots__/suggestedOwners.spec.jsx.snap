// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuggestedOwners should show owners when enabled 1`] = `
<span
  className="avatar-grid-item"
  key="1:user"
  onClick={[Function]}
  style={
    Object {
      "cursor": "pointer",
    }
  }
>
  <Tooltip
    title="<div><div class=\\"tooltip-owners-name\\"><span><span>Jane Doe</span><span> is suggested based on Issue Owner rule</span></span></div><ul class=\\"tooltip-owners-commits\\">path matched: <code>sentry/tagstore/*</code></ul></div>"
    tooltipOptions={
      Object {
        "container": "body",
        "html": true,
        "template": "<div class=\\"tooltip\\" role=\\"tooltip\\"><div class=\\"tooltip-arrow\\"></div><div class=\\"tooltip-inner tooltip-owners\\"></div></div>",
      }
    }
  >
    <ActorAvatar
      actor={
        Object {
          "id": "1",
          "name": "Jane Doe",
          "type": "user",
        }
      }
      className="tip"
      hasTooltip={false}
      title="<div><div class=\\"tooltip-owners-name\\"><span><span>Jane Doe</span><span> is suggested based on Issue Owner rule</span></span></div><ul class=\\"tooltip-owners-commits\\">path matched: <code>sentry/tagstore/*</code></ul></div>"
    >
      <Avatar
        className="tip"
        hasTooltip={false}
        title="<div><div class=\\"tooltip-owners-name\\"><span><span>Jane Doe</span><span> is suggested based on Issue Owner rule</span></span></div><ul class=\\"tooltip-owners-commits\\">path matched: <code>sentry/tagstore/*</code></ul></div>"
        user={
          Object {
            "email": "janedoe@example.com",
            "id": "1",
            "name": "Jane Doe",
          }
        }
      >
        <UserAvatar
          className="tip"
          gravatar={true}
          hasTooltip={false}
          title="<div><div class=\\"tooltip-owners-name\\"><span><span>Jane Doe</span><span> is suggested based on Issue Owner rule</span></span></div><ul class=\\"tooltip-owners-commits\\">path matched: <code>sentry/tagstore/*</code></ul></div>"
          user={
            Object {
              "email": "janedoe@example.com",
              "id": "1",
              "name": "Jane Doe",
            }
          }
        >
          <BaseAvatar
            className="tip"
            gravatar={true}
            gravatarId="janedoe@example.com"
            hasTooltip={false}
            letterId="janedoe@example.com"
            style={Object {}}
            title="Jane Doe"
            tooltip="Jane Doe (janedoe@example.com)"
            type="gravatar"
          >
            <Tooltip
              disabled={true}
              title="Jane Doe (janedoe@example.com)"
            >
              <span
                className="avatar tip"
                style={Object {}}
              >
                <img
                  onError={[Function]}
                  onLoad={[Function]}
                  src="undefined/avatar/e1f3994f2632af3d1c8c2dcc168a10e6?d=blank&s=64"
                />
              </span>
            </Tooltip>
          </BaseAvatar>
        </UserAvatar>
      </Avatar>
    </ActorAvatar>
  </Tooltip>
</span>
`;
